hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java

```diff
@@ -134,12 +161,222 @@ public boolean archiveIfRequired(HoodieEngineContext context) throws IOException
         LOG.info("No Instants to archive");
       }
 
+      if (config.getArchiveAutoMergeEnable() && !StorageSchemes.isAppendSupported(metaClient.getFs().getScheme())) {
+        mergeArchiveFilesIfNecessary(context);
+      }
       return success;
     } finally {
       close();
     }
   }
 
+  /**
+   * Here Hoodie can merge the small archive files into a new larger one.
+   * Only used for filesystem which is not supported append operation.
+   * The hole merge small archive files operation has four stages:
+   * 1. Build merge plan with merge candidates/merged file name infos.
+   * 2. Do merge.
+   * 3. Delete all the candidates.
+   * 4. Delete the merge plan.
+   * @param context HoodieEngineContext
+   * @throws IOException
+   */
+  private void mergeArchiveFilesIfNecessary(HoodieEngineContext context) throws IOException {
+    Path planPath = new Path(metaClient.getArchivePath(), mergeArchivePlanName);
+    // Flush reminded content if existed and open a new write
+    reOpenWriter();
+    // List all archive files
+    FileStatus[] fsStatuses = metaClient.getFs().globStatus(
+        new Path(metaClient.getArchivePath() + "/.commits_.archive*"));
+    // Sort files by version suffix in reverse (implies reverse chronological order)
+    Arrays.sort(fsStatuses, new HoodieArchivedTimeline.ArchiveFileVersionComparator());
+
+    int archiveFilesMergeBatch = config.getArchiveFilesMergeBatchSize();
+    long smallFileLimitBytes = config.getArchiveMergeSmallFileLimitBytes();
+
+    List<FileStatus> mergeCandidate = getMergeCandidates(smallFileLimitBytes, fsStatuses);
+
+    if (mergeCandidate.size() >= archiveFilesMergeBatch) {
+      List<String> candidateFiles = mergeCandidate.stream().map(fs -> fs.getPath().toString()).collect(Collectors.toList());
```

Comment:
Do we consider limiting the number of archive files to merge for each merging operation?  The first time triggering the archive file merging operation on an existing Hudi table with a long history could introduce the merging of a large number of archive files, penalizing a single commit.  It can also create a huge archive file, and the operation itself can take a long time and fail.  Maybe we can limit the number of files or bytes to merge?  And possibly making the merge plan to include multiple batches of merging, and each commit triggering only one batch of merging?  Sth like:

```
candidates: log1, log2, ...., log1000
plan:
batch1: log1, ..., log100 -> log1001
batch2: log101, ..., log200 -> log1002
...
batch10: log901, ..., log1000 -> log1010
this commit: only executing batch1
```

This is something we can follow up on in a separate PR.  @zhangyue19921010 you can file a ticket for that.

Reply:
Sure. https://issues.apache.org/jira/browse/HUDI-3212